finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,219 @@
+            raise DP9800Error(e)
+
+    def print_sysflag(self, sysflag: str) -> None:
+        """Print the settings of the device as stored in the system flag.
+
+        The system flag is stored as a bit mask with the format TxxLxSAF,
+        where:
+            F - bit 0: temperature unit: 0 = deg C, 1 = deg F
+            A - bit 1: audible button:   0 = silence, 1 = audible
+            S - bit 2: autoscan:         0 = no scan, 1 = autoscan active
+            x - bit 3: must be 0
+            L - bit 4: logging:          0 = no logging, 1 = logging active
+            x - bit 5: must be 0
+            x - bit 6: must be 0
+            T - bit 7: instrument type:  0 = TC, 1 = PT
+
+        Args:
+            sysflag: string representation of the system flag bitmask
+        """
+        if sysflag != "":
+            instr_type = ["TC", "PT"][int(sysflag[0])]
+            logging_state = ["no logging", "logging active"][int(sysflag[3])]
+            scanning_state = ["no scan", "autoscan active"][int(sysflag[5])]
+            audible_state = ["silence", "audible"][int(sysflag[6])]
+            temp_unit = ["deg C", "deg F"][int(sysflag[7])]
+            print(f"Instrument type: {instr_type}")
+            print(f"Logging: {logging_state}")
+            print(f"Autoscan: {scanning_state}")
+            print(f"Audible button: {audible_state}")
+            print(f"Temperature unit: {temp_unit}")
+
+    def read(self) -> bytes:
+        """Read temperature data from the DP9800.
+
+        The DP9800 returns a sequence of bytes containing the
+        temperatures measured on each channel, in the format
+
+          STX T SP t1 SP t2 SP t3 SP t4 SP t5 SP t6 SP t7 SP t8 SP t9 ff ETX BCC NUL
+
+        where
+            t1, t2, ..., t9: temperature values in the format %4.2f
+            STX: Start of Text (ASCII 2)
+            ETX: End of Text (ASCII 3)
+            NUL: Null character (ASCII 0)
+            SP: Space (ASCII 32)
+            BCC: Block Check Character
+            ff: System flag in hexadecimal
+
+        For error checking, the BCC is calculated by performing consecutive XOR
+        operations on the message and compared with the BCC received.
+
+        Note a slight peculiarity:
+        The device actually returns 9 temperatures, while the documentation
+        states that there should be 8. The device shows 8 values, corresponding
+        to the values with indices 1 to 8. The first value (at index 0) is
+        therefore ignored.
+
+        Returns:
+            data: the sequence of bytes read from the device
+
+        Raises:
+            DP9800Error: Malformed message received from device
+        """
+        if self.serial.in_waiting == 0:
+            raise DP9800Error("No data waiting to be read")
+
+        try:
+            data = self.serial.read_until(b"\x00")
+        except SerialTimeoutException:
+            raise DP9800Error("Read request timed out")
+
+        return data
```

Comment:
Unfortunately, `read_until()` *doesn't* raise an exception if it times out; it just returns fewer bytes than expected (usually zero bytes): https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.read_until

I think you want to set the `size` argument to 4 (or something close). (I'm thinking you need 3 bytes for the header and one for the terminator.) Then you need to check whether `len(data)` is big enough and raise an exception if not (which I'd do in this function rather than your parsing one).

`read_until()` *might* throw a `SerialException`, though, so that's the one you need to check for.

Reply:
I don't think we should be hardcoding that though. I'd just make `size` be the minimum required for the code that comes after to work, i.e. the size of the header plus checksum plus terminator. Otherwise if someone wanted to change the device for one that returns, say, six or ten values instead, they'd have to figure out that they also had to change this one random magic number.

Reply:
@alexdewar , do not think that much on possible use cases. This class is specific for DP9800, which returns 9 temperature values no matter what, so let's write code design to work with that. If this were a base class, or a generic interface, it would be different, but this is bespoken to a particular device.

Reply:
So, the dummy should be generic, but not this one.

Reply:
Yeah, I don't think I expressed myself very well there. I'm pretty sure that device will never be replaced.

Originally the way @dc2917 was extracting the temperatures assumed there were nine channels (which is fine because there are!) but I suggested it would be a bit cleaner to just `split()` it rather than hardcoding a bunch of offsets etc. So this suggestion about the number of bytes is in a similar vein.